### DISCUSSION TITLE
Why Am i getting this error .map is not a function

### DISCUSSION THREAD

**Discussion by mandorakannu** [vercel/next.js#49990] [Help]
### Summary

After doing server side rendering, sometimes i got the error and i dont know how to fix it.

`Blogs/page.tsx` `Blog fetch function`

```
 async function getAllBlogs() {
  try {
    const url = process.env.BASE_URL;
    const res = await fetch(url,{
      cache: "no-cache",
      next: {revalidate: 10}
    });
    const data = await res.json();
    return data;
  } catch (error: unknown) {
    console.log(error);
  }
};
```

`Main react component`
```
async function Blogs(): Promise<JSX.Element> {
  const blogs: Blogs = await getAllBlogs();
  return (
    <>
      <section
        className={`grid max-sm:grid-cols-1 md: grid-cols-2 md:grid-cols-3 gap-6 m-6 cursor-default ${outfit.className}`}
      >
        {blogs.map((blog) => (
          <div
            key={blog._id}
            className="flex flex-col justify-between p-4 bg-white rounded-lg shadow-lg border"
          >
              <div className="flex flex-col justify-start gap-2">
                <h1 className="text-xl font-bold">{blog.title}</h1>
                <h2>{blog.subTitle}</h2>
              </div>
              <Link
                href={`/Blogs/${blog._id}`}
                className="text-red-500 my-3 w-fit hover:underline underline-offset-4"
              >
                Read More...
              </Link>
          </div>
        ))}
      </section>
    </>
  );
}
```

### Additional information

```js
But I got an error but its gone If i refresh page 5-6 times then all blogs is listed.


- error app\(common)\Blogs\page.tsx (44:15) @ map
- error TypeError: Cannot read properties of undefined (reading 'map')
    at Blogs (Blogs/page.tsx:35:29)
  42 |         className={`grid max-sm:grid-cols-1 md: grid-cols-2 md:grid-cols-3 gap-6 m-6 cursor-default ${outfit.className}`}
  43 |       >
> 44 |         {blogs.map((blog) => (
     |               ^
  45 |           <div
  46 |             key={blog._id}
  47 |             className="flex flex-col justify-between p-4 bg-white rounded-lg shadow-lg border"
    syscall: 'getaddrinfo',
    hostname: 'http'
  }
}
```
```


### Example

https://mandorakannu.me/Blogs

**Comment by devjiwonchoi** [comment]
You are using `cache: "no-cache",
      next: {revalidate: 10}` together which causes an error when revalidating.
`revalidate: 10` is identical to `cache: force-cache`and will be conflicted with `cache: "no-cache"` which will cause an error.

In your case, it is commented directly in the [docs](https://nextjs.org/docs/app/api-reference/functions/fetch)

> Conflicting options such as { revalidate: 0, cache: 'force-cache' } or { revalidate: 10, cache: 'no-store' } will cause an error.

To resolve this issue, choose either `cache: no-store` or `revalidate: 10` based on your needs, but do not use them together.

P.S. "no-cache" tag and "no-store" are [functionally identical](https://nextjs.org/docs/app/api-reference/functions/fetch). 

> The no-cache option behaves the same way as no-store in Next.js.

**Reply by mandorakannu** [reply]
Nah Bro, This is not working

**Reply by devjiwonchoi** [reply]
Try to return an empty string if catch an error.
If fetch has been failed for API issue, your page has nothing to `.map` since it is possibly returning undefined from `blogs`.

Example:
```
async function getAllBlogs() {
  try {
    const url = process.env.BASE_URL;
    const res = await fetch(url,{
      next: { revalidate: 10 }
    });
    const data = await res.json();
    return data;
  } catch (error: unknown) {
    console.log(error);
    return [];
  }
}
```

**Comment by icyJoseph** [comment]
Is it logging any errors? Are you sure your function returns a list? 

This line:

```console
error TypeError: Cannot read properties of undefined (reading 'map')
```

Tells me that `getAllBlogs` is not returning what you expect. The proper way to attack this problem is to either debug, or log returned values. Most likely the catch branch of your try/catch is actually kicking in, but who knows why... 

It might also be that the:

```js
 const res = await fetch(url,{
      next: { revalidate: 10 }
    });
```

is returning 4xx, and because fetch doesn't throw on those, you continue onto an error. Maybe add this bit:

```js
 const res = await fetch(url,{
      next: { revalidate: 10 }
    });
if (!res.ok) {
   throw new Error(`failed to fetch ${res.status}`);
}
```

And from there, you should be able to add more supporting code, or fix the fetch. It might be that, even though the response is not OK, you can still parse it as json, and get helpful data from it, etc...

In general, you have to check all of your assumptions right now. Can you even get data by just making a request to `process.env.BASE_URL` using some HTTP client like POSTMAN, browser fetch or [VSCode REST Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), or [ThunderClient](https://marketplace.visualstudio.com/items?itemName=rangav.vscode-thunder-client), curl, etc?

**Comment by SkyeFull90** [comment]
I get this same error to multiple times especially when using mongoose and all my apis work but the .map function constantly breaks if i change anything

**Reply by icyJoseph** [reply]
What error do you get? Whats the type of the data you get? `typeof data`?

**Reply by icyJoseph** [reply]
@SkyeFull90 I found this in your repository: https://github.com/SkyeFull90/skyeServiceDir/blob/main/projects/mongodbcrud/app/compoents/EventsList.jsx

There I see:

- Error not awaiting:

```js
  const { topic } = getTopics();
 // should be
  const { topic } = await getTopics();
```

- Topics API route won't be present during build time:

```js
  const res = await fetch('http://localhost:3000/api/topics', {
      cache: 'no-store'
    })
```

Which points to, https://github.com/SkyeFull90/skyeServiceDir/blob/main/projects/mongodbcrud/app/api/topics/route.js, where all you do is connect to a DB to fetch data. You should move that into https://github.com/SkyeFull90/skyeServiceDir/blob/main/projects/mongodbcrud/app/compoents/EventsList.jsx -> `getTopics`:

```js
// pseudo-code
import connectMongodb from "@/libs/mongodb";
import Topic from "@/models/topic";
// other imports

const getTopics = async()=>{
  try {
    await connectMongodb();
    const topics = await Topic.find();

    return topics;
  } catch (error) {
    console.log("Error loading events", error)
  }
}

```

There's no server to respond to your API request during build time. And even if there were a server, it is unnecessary. This comes from [Pages Router](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props#getserversideprops-or-api-routes) but it is just as true in the App Router, it is why the App Router exists:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/pages/building-your-application/routing/api-routes) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

Take this example, https://github.com/rauchg/how-is-this-not-illegal - no need to make more API calls (motivated by a false sense of modularity or concern separation), because the Server Components, can and should fetch the data directly.

**Reply by SkyeFull90** [reply]
Thank you for this it fully compiled and built. After I did the refactoring you said to do. I appreciate the help.
